### PR TITLE
[READY] Silence MSVC warnings

### PR DESCRIPTION
**Pull request by bstaletic**
Abseil produces ~40'000 lines of warnings when ycmd is compiled on Windows. This should silence them.

---
This change is [Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1566)

### CODE REVIEW

**Comment by bstaletic** [comment]
Aaaaand it didn't work at all.

**Review by puremourning — COMMENTED** [review]
should we be passing `/std:c++17`  to msvc too?

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1566#-:-MZU2RQz-SOlN0rnlPXT:b-92lzm6)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic)

**Review by puremourning — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1566)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic)

___
*[cpp/CMakeLists.txt, line 187 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1566#-MZU3Huk6agnhGscqHLb:-MZU3Huk6agnhGscqHLc:b-cf4q7r) ([raw file](https://github.com/ycm-core/ycmd/blob/114641ebd95c38431c92215a9422965e131110ae/cpp/CMakeLists.txt#L187)):*
> ```CMake
> # Makes MSVC conform to the standard
> if ( MSVC )
>   set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /permissive- /Zc:__cplusplus /Zc:externConstexpr /Zc:inline /Zc:preprocessor /Zc:throwingNew" )
> ```

like wow. reading the msvc docs for this is like pain after pain. They are like "enable standard conformant behaviour, because our old default behaviour was just bad, or pre standardisation and we don't want to break everyone by "fixing" it "

**Review by bstaletic — COMMENTED** [review]
> should we be passing /std:c++17  to msvc too?

No, that's taken care of a bit further below where we set `CMAKE_CXX_STANDARD.

&nbsp;

There are some more warnings, but I have no idea how do fix them while staying portable (other than ugly preprocessor magic).

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1566#-:-MZaVOSk7kly7yXjyDju:b-8r79fq)* status: 1 of 2 LGTMs obtained (waiting on @bstaletic)

___
*[cpp/CMakeLists.txt, line 187 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1566#-MZU3Huk6agnhGscqHLb:-MZaVbE-0S46NO7n54e6:b-ict0fl) ([raw file](https://github.com/ycm-core/ycmd/blob/114641ebd95c38431c92215a9422965e131110ae/cpp/CMakeLists.txt#L187)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

like wow. reading the msvc docs for this is like pain after pain. They are like "enable standard conformant behaviour, because our old default behaviour was just bad, or pre standardisation and we don't want to break everyone by "fixing" it "
</blockquote></details>

So... here's a quick rehash of what each flag does:

1. `/permissive-` enforces two-phase template instantiation (among other things) and VS 2017 can't compile pybind11 in C++17 mode without it. We absolutely need it.
2. Without `/Zc:__cplusplus` MSVC goes "mhmm... `#define __cplusplus 199711L` looks about right." Then it defines `_MSVC_LANG` to the correct value of `__cplusplus`. This was causing 40k lines of warnings from abseil.
3. `/Zc:externConstexpr` fixes linkage of constexpr variables.
4. `/Zc:inline` is supposed to remove unreferenced `inline` functions.
5. `/Zc:throwingNew` says "operator new doesn't return NULLs".

In other words, MSVC isn't a C++ compiler. It is a compiler that compiler something that only quite resembles C++.

**Review by puremourning — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1566#-:-MZaydSPBts8byVYFAAi:bnfp4nl)* status: 1 of 2 LGTMs obtained (waiting on @bstaletic)
